ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -33,23 +34,43 @@ def __init__(self):
 
         rospy.Subscriber('/current_pose', PoseStamped, self.pose_cb)
         rospy.Subscriber('/base_waypoints', Lane, self.waypoints_cb)
+        rospy.Subscriber('/traffic_waypoint', Int32, self.traffic_cb)
+        rospy.Subscriber('/obstacle_waypoint', Lane, self.obstacle_cb)
 
-        # TODO: Add a subscriber for /traffic_waypoint and /obstacle_waypoint
-
-        self.final_waypoints_pub = rospy.Publisher('final_waypoints', Lane,
-                                                   queue_size=1)
+        self.pub = rospy.Publisher('final_waypoints', Lane, queue_size=1)
 
         # TODO: Add other member variables you need below
+        self.world = None
+        self.ego = None
 
+        self.publish()
```

Comment:
This should not be here! First because `publish` is an infinite loop the way you have implemented, so it will never reach `spin`. And it's very very important that `spin` executes (internally it's an infinite loop as well).

What `spin` does is to indefinitely keep polling the `subscribers` to see if they got some data. If they do, the corresponding callback function will be called.
The `publish` part should be in some callback function - once you get the data that you need, you process it and you publish. I believe the best place is the callback to `/current_pose` since it has the fastest rate.

So you don't even need to define a publish rate or create a while loop, just publish whenever a new message from `/current_pose` arrives.

Reply:
I also thought of doing that, however I wanted to follow what the waypoint loader does and it seems that it never reaches `rospy.spin()` too.

Reply:
Yes, that's true. I believe it's a bit misleading that they use `rospy.spin()`.
The code in `waypoint_loader.py` works because it doesn't have any subscribers, thus `spin()` is unnecessary (have a look here: http://wiki.ros.org/ROS/Tutorials/WritingPublisherSubscriber%28c%2B%2B%29#roscpp_tutorials.2BAC8-Tutorials.2BAC8-WritingPublisherSubscriber.CA-77f59f9b7961ee481f7596923b938df8c19148f6_103)

Another approach is to have the `while` loop and put inside a `rospy.spinOnce`, but I think it's more intuitive to publish right after receiving data.